pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -31,57 +34,125 @@ import (
 var (
 	k8sOrchestratorInstance         *K8sOrchestrator
 	onceFork8sOrchestratorInstance  sync.Once
-	featureStatesConfigMapName      string
-	featureStatesConfigMapNamespace string
+	supervisorFSSConfigMapName      string
+	supervisorFSSConfigMapNamespace string
+	internalFSSConfigMapName        string
+	internalFSSConfigMapNamespace   string
+	clusterFlavor                   cnstypes.CnsClusterFlavor
 )
 
 // K8sOrchestrator defines set of properties specific to K8s
 type K8sOrchestrator struct {
-	featureStates   map[string]string
-	informerManager *k8s.InformerManager
+	supervisorFeatureStates map[string]string
+	internalFeatureStates   map[string]string
+	informerManager         *k8s.InformerManager
+}
+
+// K8sInitParams lists the set of parameters required to run the init for K8sOrchestrator
+type K8sInitParams struct {
+	InternalFeatureStatesConfigInfo   config.FeatureStatesConfigInfo
+	SupervisorFeatureStatesConfigInfo config.FeatureStatesConfigInfo
 }
 
 // Newk8sOrchestrator instantiates K8sOrchestrator object and returns this object
-func Newk8sOrchestrator(ctx context.Context, featureStatesConfigMapInfo config.FeatureStatesConfigInfo) (*K8sOrchestrator, error) {
-	var coInstanceErr error
+// NOTE: As Newk8sOrchestrator is created in the init of the driver and syncer components,
+// raise an error only if it is of utmost importance
+func Newk8sOrchestrator(ctx context.Context, controllerClusterFlavor cnstypes.CnsClusterFlavor, params K8sInitParams) (*K8sOrchestrator, error) {
+	var (
+		coInstanceErr error
```

Comment:
Since this is a local variable to the function, it will be redeclared for each function call. Since `onceFork8sOrchestratorInstance` is a singleton, i think the expectation is that if the first call fails, we always return the same error. However in this case, we're going to return `nil, nil` on line 86 and i'm not sure how callers handle that..

Reply:
We do not retry creating a k8sOrchestrator if we get a failure. But, I agree with you about the `nil, nil` part. 
@chethanv28 I did not change the logic for this error return, could you reply on Raunak's question instead?

Reply:
Kindly refer to https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/419#discussion_r505765241